zcash_client_sqlite/src/wallet/sapling.rs

```diff
@@ -136,37 +136,51 @@ fn to_spendable_note<P: consensus::Parameters>(
             SqliteClientError::CorruptedData("Note commitment tree position invalid.".to_string())
         })?);
 
-    let ufvk_str: String = row.get(7)?;
-    let ufvk = UnifiedFullViewingKey::decode(params, &ufvk_str)
-        .map_err(SqliteClientError::CorruptedData)?;
-
-    let scope_code: i64 = row.get(8)?;
-    let spending_key_scope = parse_scope(scope_code).ok_or_else(|| {
-        SqliteClientError::CorruptedData(format!("Invalid key scope code {}", scope_code))
-    })?;
-
-    let recipient = match spending_key_scope {
-        Scope::Internal => ufvk
-            .sapling()
-            .and_then(|dfvk| dfvk.diversified_change_address(diversifier)),
-        Scope::External => ufvk
-            .sapling()
-            .and_then(|dfvk| dfvk.diversified_address(diversifier)),
-    }
-    .ok_or_else(|| SqliteClientError::CorruptedData("Diversifier invalid.".to_owned()))?;
-
-    Ok(ReceivedNote::from_parts(
-        note_id,
-        txid,
-        output_index,
-        Note::Sapling(sapling::Note::from_parts(
-            recipient,
-            sapling::value::NoteValue::from_raw(note_value),
-            rseed,
-        )),
-        spending_key_scope,
-        note_commitment_tree_position,
-    ))
+    let seed_fingerprint = row
+        .get::<_, Option<[u8; 32]>>(7)?
+        .map(HdSeedFingerprint::from_bytes);
+    let ufvk_str: Option<String> = row.get(8)?;
+    let scope_code: Option<i64> = row.get(9)?;
+
+    // If we don't have information about the recipient key scope or the seed fingerprint, we can't
+    // determine which spending key to use. This may be because the received note was associated
+    // with an imported viewing key, so we treat such notes as not spendable.
```

Comment:
Hmm: at the moment, _we_ can't determine that, but in theory the seed phrase holder might have a mapping from `InputSelector::AccountId` to USK. `create_proposed_transactions` only takes a USK, not a seed, so we can't usefully use the seed fingerprint.

However, when we eventually _are_ the Rust wallet (as part of `zcashd` deprecation), we will definitely need such a mapping ourselves, which we can use the seed FP for (caller will fetch the seed FP and ZIP 32 account index, and use those to determine which USK to derive from their seed). So I can see the rationale that for simplicity we treat imported UFVKs as view-only. If this is the direction we take, then when UFVK import is exposed we need to document it as such.

Reply:
In the future, I think what we should probably do is take the seed fingerprint(s) as an argument when we go to initially request spendable inputs, so that we know what seed(s) we'll be spending from?

Reply:
I'm perplexed at the thought that a seed is assumed to have even been seed by the database in order to spend funds. An imported account won't provide a seed or fingerprint, yet it can absolutely be spendable. Rather than use the fingerprint and zip32 account index to ask for the USK, could we instead use the UFVK as the 'key' by which we ask for the USK?

Reply:
UFVKs are not suitable for use as "map keys", due to being containers themselves. Once arbitrary UFVK import is allowed, we have no way of enforcing that its component FVKs are at all tied to the same ZIP 32 account or seed phrase. And when additional receivers are added in future, or metadata is added to a UFVK, the matching becomes more complex. In general, the result is that instead of using the UFVK as a key, you would need to do an arbitrary search for seed or key material that can reproduce each of the known component FVKs within a given UFVK, which is not great UX.

Reply:
But as I said in my original comment, as long as we know the UFVK for an account (so we are acting as a wallet), and the surrounding app knows all spending key material for that UFVK, then we can create transactions. (Conversely, we _cannot_ ever do this for UIVKs, because we cannot track balance, so we cannot know whether any given note is spent or not.)

So if we can figure out an approach with reasonable UX for mapping arbitrary imported UFVKs to their key material, then we could support that. But I don't think we have an answer for this at the moment (at least, I do not have time to think of one on the Orchard implementation timeline)

Reply:
I've removed the seed fingerprint check.